tests/integration/pkg/main_test.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"sigs.k8s.io/e2e-framework/pkg/env"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/support/kind"
+)
+
+const (
+	envNameUseKubeConfig = "USE_KUBECONFIG"
+	envNameKubeConfig    = "KUBECONFIG"
+)
+
+var (
+	testenv env.Environment
+)
+
+func TestMain(m *testing.M) {
+	if len(os.Getenv(envNameUseKubeConfig)) > 0 {
+		testenv = env.NewWithKubeConfig(os.Getenv(envNameKubeConfig))
+	} else {
+		testenv = env.New()
+		kindClusterName := generateID("cluster")
+
+		// Use pre-defined environment funcs to create a kind cluster prior to test run
+		testenv.Setup(
+			CreateKindCluster(kindClusterName),
+		)
+
+		// Use pre-defined environment funcs to teardown kind cluster after tests
+		testenv.Finish(
+			DestroyKindCluster(kindClusterName),
+		)
+	}
+
+	os.Exit(testenv.Run(m))
+}
+
+type kindContextKey string
+
+func CreateKindCluster(clusterName string) env.Func {
+	return func(ctx context.Context, cfg *envconf.Config) (context.Context, error) {
+		k := kind.NewCluster(clusterName)
+		// We only provide the config because the API is constructed in such a way
+		// that it requires both the image and the cluster config.
+		kubecfg, err := k.CreateWithConfig(KindImages.Default, "yamls/cluster.yaml")
+		if err != nil {
+			return ctx, err
+		}
+
+		// TODO: find a better way than sleeping here, e.g. to wait for specific
+		// k8s subsystems to be avialable.
```

Comment:
Would doing `kubectl version --request-timeout 3s` work here? Or are we waiting for something specific?

Reply:
We could but we don't have kubectl options readily available here and we'd need to create them (nothing huge but it would need to be done).

+ I'd prefer to add more logic to this "wait" so that we wait e.g. for all subcomponents/daemonsets/deployments(like `coredns`) to be up and running.

Reply:
How about: `kubectl wait --for=condition=Ready pods --all --all-namespaces --timeout 3s`? That waits until all Pods are ready, which is a pretty good approximation of what we want here, I think. I don't really think we can avoid setting up kubectl here if we want to interact with the cluster meaningfully.

FYI I'm fine merging this as is, I just thought there might be a simple solution.